src/GraphWebhooks/Controllers/ListenController.cs

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using GraphWebhooks.Models;
+using GraphWebhooks.Services;
+using GraphWebhooks.SignalR;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Graph;
+using Microsoft.Identity.Web;
+
+namespace GraphWebhooks.Controllers
+{
+    /// <summary>
+    /// Implements the notification endpoint which receives
+    /// notifications from Microsoft Graph
+    /// </summary>
+    public class ListenController : Controller
+    {
+        private readonly GraphServiceClient _graphClient;
+        private readonly SubscriptionStore _subscriptionStore;
+        private readonly CertificateService _certificateService;
+        private readonly IHubContext<NotificationHub> _hubContext;
+        private readonly ILogger<ListenController> _logger;
+        private readonly List<Guid> _appIds;
+        private readonly List<Guid> _tenantIds;
+
+        public ListenController(
+            GraphServiceClient graphClient,
+            SubscriptionStore subscriptionStore,
+            CertificateService certificateService,
+            IHubContext<NotificationHub> hubContext,
+            IConfiguration configuration,
+            ILogger<ListenController> logger)
+        {
+            _graphClient = graphClient;
```

Comment:
defensive programming

Reply:
Are null checks required for DI? I've looked into this before and none of our docs address it, but all of the example code concerning it just assigns without any checking.

Reply:
by definition a class that get's things from DI, doesn't know it gets things from a DI framework.
if most DI framework out there fail if they don't have enough information corresponding to a dependent's constructor, it's not always necessary the case.
Plus that class could perfectly be manually newed up as well, for unit testing purposes for examples.
For all reasons, a class, whether it's used in DI contexts or not, should defend itself against wrong external usage (in this case passing null values that are going to make the class fail later on).